The comment beginning about `black` makes me feel conflicted... More on that below.

@eli-schwartz yes I strongly agree, programmers must learn the languages and tools they're using. At every opportunity, I'm pushing my colleagues and collaborators to learn to understand what they're doing, better. (And vice versa, willing to receive that.) However, in my experience such "help" is commonly unwanted; all too often, people don't want to learn — and best of all, not always for bad reasons.

In this regard, @stevenbenner has handled the PR feedback very well! Please don't delay review/merge; doing that with this discussion is not my intent. I'm just missing out on genuine technical open discussion, haven't had that in years.

-------------------------------

> [shellcheck is] great for people who simply do not know bash at all

I find this a too-easy judgement to make seriously :smiling_face_with_tear: 

I'd written a great chunk of bash, zsh, posix sh, nowadays even `fish`. One "masterpiece" I did in bash had GUI and did IPC of sorts between several background processes (coprocesses, in bash terms). I've learned enough of bash to know to not fool myself that I'll get every single line right on the 1st try. Then, perusing shellcheck hints simply saves time, as a quick automatic checklist.

I'm not attacking bash, no way! It's a good time-proven tool, with great power, and sure enough, great trickiness :sweat_smile:

I see no practical issue with ignoring the lints. Doing it explicitly (i.e. `# shellcheck disable=SC2086`) has the slight benefit of expressing "the author has reviewed this, and it was deemed useless" — which preempts unwanted suggestions like me and @alerque have made here.

Look, there's immense value to the cognitive tool of abstraction. Without abstraction, every coder would have to learn all the way down to assembly instructions, if not to cpu transistor level, right? But viewed another way, abstraction is nothing but willful ignorance: one just assumes a simple model, and chooses to ignore at-the-moment irrelevant details.

Individually, I'm opposing the practice of cargo-culted enforcement of linter compliance (when CI fails unless zero lints). Coding against that feels like programming from jail. Yet, some "people in programmer roles" don't care so egregiously much, that I'd introduced those exact jails myself, to slow down deterioration of codebases.

Likewise, individually I'm opposing the `black` formatter for python. It shuts down freedom of expression; it drains decent code dry of life and joy. Turns code writing, and reading, into robotic bureaucracy. Yet! When an immature team fails to get along, and spirals into style-nitpicking wars in code-review, and a manager fails to resolve this on human level — introducing mandatory uncompromising `black` style will be a wrong and incomplete, but a solution.

> [...] Static analyzers are kind of doomed here.

Agree; and I'll further add that's true of any (turing-complete) language, even those without eval. It's a mathematical fact from computability theory, see [Rice's theorem](https://en.wikipedia.org/wiki/Rice%27s_theorem), its statement is fairly simple. Yet, despite the grim theoretical prognosis in the most general case — restricting oneself to special cases enables having useful analyzers anyway.

----------------

> Yes. I know that. You're responding to my advice which pointed out the difference... by informing me that there's a difference?

Yes... you're right. I apologize for the confusion and wasted time.
Also, sincere gratitude for responding and the discussion.

--- a/aurpublish.in
+++ b/aurpublish.in
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+REMOTE_URL='aur@aur.archlinux.org:'
+
 #### Declare common functions
 
 # usage and other awkward blocks of preformatted text
@@ -35,9 +37,8 @@ _EOF_
 ssh_config() {
 	cat <<- _EOF_
 		\`\`\`
-		Host aur aur.archlinux.org
+		Host aur.archlinux.org
 		    User aur
-		    Hostname aur.archlinux.org
 		    IdentityFile ~/.ssh/keys/aur
 		\`\`\`
 _EOF_
@@ -142,9 +143,9 @@ if (( PULL_SUBTREE )); then
     # test if prefix already exists
     if is_package_in_git ${package}; then
         git subtree split -P "${package}" --rejoin
-        git subtree pull -P "${package}" aur:${pkgbase}.git master -m "Merge subtree '${package}'"
+        git subtree pull -P "${package}" ${REMOTE_URL}/${pkgbase}.git master -m "Merge subtree '${package}'"
     else
-        git subtree add -P "${package}" aur:${package}.git master
+        git subtree add -P "${package}" ${REMOTE_URL}/${package}.git master
     fi
     exit 0
 fi
@@ -152,4 +153,4 @@ fi
 if (( SPEEDUP )); then
     git subtree split -P "${package}" --rejoin
 fi
-git subtree push -P "${package}" aur:${pkgbase}.git master
+git subtree push -P "${package}" ${REMOTE_URL}/${pkgbase}.git master

--- a/aurpublish.in
+++ b/aurpublish.in
@@ -147,6 +147,18 @@ fi
 if (( ! URL_SET )); then
     # use git configured remoteUrl, if set
     REMOTE_URL=$(git config --default "${REMOTE_URL}" --get aurpublish.remoteUrl | sed 's:/*$::')
+
+    # warn the user if they have a non-default ssh alias for 'aur'
+    auralias=$(ssh -G aur | awk '$1 == "hostname" { print $2 }')
+    if [[ ${auralias} != 'aur' && ${auralias} != 'aur.archlinux.org' ]]; then
+        echo "Warning: You have a non-default alias for 'aur' in your ssh configuration. (${auralias})"
+        echo "Previous versions of aurpublish used the 'aur' ssh hostname, but you should now use the --url option."
+        echo ""
+        echo "This operation will use the git server at: ${REMOTE_URL}"
+        echo ""
+        read -p "Do you wish to continue? [y/n] " userwarned
+        [[ ${userwarned} = [yY] ]] || exit 1
+    fi
 fi
 
 pkgbase="$(sed -rn 's/pkgbase = (.*)/\1/p' "${package}"/.SRCINFO 2>/dev/null)"

--- a/aurpublish.in
+++ b/aurpublish.in
@@ -22,6 +22,8 @@ usage() {
 		                      and a second copy of all commits in the subtree.
 		                      For more details, see the "--rejoin" option in git
 		                      subtree.
+		    -u, --url       Specify the URL of the server that should be used
+		                      for git operations.
 		    -h, --help      Show this usage message
 
 		COMMANDS
@@ -69,6 +71,11 @@ while [[ "${1}" != "" ]]; do
             ;;
         -s|--speedup)
             SPEEDUP=1
+            ;;
+        -u|--url)
+            REMOTE_URL=${2%/}
+            URL_SET=1
+            shift
             ;;
         log)
             LOG=1
@@ -135,6 +142,11 @@ if (( LOG )); then
     shift
     git log "${log_pre_opts[@]}" $(git subtree split -P "${package}") "$@"
     exit 0
+fi
+
+if (( ! URL_SET )); then
+    # use git configured remoteUrl, if set
+    REMOTE_URL=$(git config --default "${REMOTE_URL}" --get aurpublish.remoteUrl | sed 's:/*$::')
 fi
 
 pkgbase="$(sed -rn 's/pkgbase = (.*)/\1/p' "${package}"/.SRCINFO 2>/dev/null)"

--- a/doc/aurpublish.1.asciidoc
+++ b/doc/aurpublish.1.asciidoc
@@ -49,6 +49,9 @@ Options
         commits in the subtree. For more details, see the "--rejoin"
         option in linkman:git-subtree[1].
 
+*-u*, *--url* <URL>::
+        Specify the URL of the server that should be used for git operations.
+
 *-h*, *--help*::
         Prints a usage page.
 
@@ -62,6 +65,29 @@ Hooks
 *prepare-commit-msg*::
         Prefill the commit message with a list of added/updated/deleted
         packages + versions (if any).
+
+
+Specifying AUR server URL
+-------------------------
+You can use aurpublish to interact with a git server other than the default
+AUR git server at aur.archlinux.org. This may be useful if you are running
+your own AUR server.
+
+The upstream git server URL can be configured in two ways. You can set the
+URL with the -u/--url options when you run aurpublish. Or you can configure
+the server persistently in git by setting the aurpublish.remoteUrl option.
+
+Set the URL with --url::
+        aurpublish --url ssh://me@example.com mypackage
+
+Set the URL for the current git directory with git config::
+        git config aurpublish.remoteUrl ssh://me@example.com
+
+Set the URL globally with git config::
+        git config --global aurpublish.remoteUrl ssh://me@example.com
+
+For more details on git options see linkman:git-config[1].
+
 
 Examples
 --------
@@ -79,6 +105,9 @@ aurpublish ansible-core-git::
 
 aurpublish log::
         View the git log of a package subtree.
+
+aurpublish --url ssh://me@example.com ansible-core-git::
+        Push the subtree to a server at ssh://example.com.
 
 AUTHORS
 -------

